src/mempool/block_receiver.rs

```diff
@@ -0,0 +1,201 @@
+                        None,
+                    ))
+                    .await
+                    .unwrap();
+            }
+        }
+    }
+
+    async fn sync_missing_blocks(&mut self, start_block_number: u64, stop_block_number: u64) {
+        info!(
+            start_block_number,
+            stop_block_number, "Syncing missing blocks",
+        );
+        let mut current_block_number = start_block_number;
+        while current_block_number <= stop_block_number {
+            let (block_tx, block_rx) = oneshot::channel::<Option<Block>>();
+            self.system_tx
+                .send(SystemMessage::BlockRequest {
+                    block_number: current_block_number,
+                    block_tx,
+                })
+                .await
+                .unwrap();
+            let block = block_rx.await.unwrap().unwrap();
+            if block.events.len() == 0 {
+                current_block_number += 1;
+                continue;
+            }
+            self.submit_block(&block).await;
+            current_block_number += 1;
+        }
+    }
+
+    pub async fn run(&mut self) {
+        info!(shard = self.shard_id.to_string(), "Running block receiver");
+        loop {
+            let block = self.block_rx.recv().await.unwrap();
+            info!(
+                shard = self.shard_id.to_string(),
+                num_events = block.events.len(),
+                height = block.header.as_ref().unwrap().height.unwrap().block_number,
+                "Received block"
+            );
+            if block.events.len() == 0 {
+                continue;
+            }
+            // The db is the source of truth, it's possible to read this out of the events_rx channel but delivery over that channel is not reliable (it's a broadcast channel) we may not have the most up to date state.
+            let last_stored_event_seqnum = self.stores.block_event_store.max_seqnum().unwrap();
+            let last_event_in_block = block.events.last().unwrap();
+            if last_event_in_block.seqnum() < last_stored_event_seqnum {
+                continue;
+            }
+
+            let first_event_in_block = block.events.first().unwrap();
+            if first_event_in_block.seqnum() > last_stored_event_seqnum + 1 {
+                let last_stored_event = self
+                    .stores
+                    .block_event_store
+                    .get_last_block_event()
+                    .unwrap();
+                let last_block_number = match &last_stored_event {
+                    None => 1,
+                    Some(last_event) => last_event.block_number(),
+                };
+                self.sync_missing_blocks(
+                    last_block_number,
+                    block.header.as_ref().unwrap().height.unwrap().block_number - 1,
+                )
+                .await;
+                if let Err(BlockReceiverError::ConfirmationTimedOut) = self
+                    .wait_for_confirmation(
```

Comment:
I'm not sure this is safe? If we only call recv on events_rx periodically, then the events broadcast channel could fill up and cause us to miss events (and potentially interfere with other subscribers? I need to read up on the implementation).

I think we may need to be constantly listening to events messages in the background and keep tracking the max seq num per shard.

Reply:
I think this behavior works. It won't interfere with other subscribers. The way the channel works is that it keeps the last n messages and if there's a slow subscriber it drops the oldest message in the buffer and that subscriber will miss events. 

These events should be produced at a very regular 1s cadence so it's fine if some events are dropped because the next one will show up with the latest event seqnum.